src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,631 @@
+     * If the required parameters were not supplied and can be generated by
+     * the {@code KDF} object, the generated parameters are returned;
+     * otherwise {@code null} is returned.
+     *
+     * @return the parameters used with this {@code KDF} object, or
+     * {@code null}
+     *
+     * @throws UnsupportedOperationException if this method is not overridden
+     * by a provider
+     */
+    public KDFParameters getKDFParameters() {
+        chooseFirstProvider();
+        return spi.engineGetKDFParameters();
+    }
+
+    /**
+     * Returns a {@code KDF} object that implements the specified algorithm.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use.
+     *     See the {@code KDF} section in the <a href=
+     *     "{@docRoot}/../specs/security/standard-names.html#kdf-algorithms">
+     *     Java Security Standard Algorithm Names Specification</a>
+     *     for information about standard KDF algorithm names.
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if no {@code Provider} supports a {@code KDF} implementation for the
+     *     specified algorithm
+     * @throws NullPointerException
+     *     if {@code algorithm} is {@code null}
+     */
+    public static KDF getInstance(String algorithm)
+        throws NoSuchAlgorithmException {
+        try {
+            return getInstance(algorithm, (KDFParameters) null);
+        } catch (InvalidAlgorithmParameterException e) {
+            throw new NoSuchAlgorithmException(
+                "Received an InvalidAlgorithmParameterException. Does this "
+                + "algorithm require KDFParameters?", e);
+        }
+    }
+
+    /**
+     * Returns a {@code KDF} object that implements the specified algorithm from
+     * the specified security provider.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use.
+     *     See the {@code KDF} section in the <a href=
+     *     "{@docRoot}/../specs/security/standard-names.html#kdf-algorithms">
+     *     Java Security Standard Algorithm Names Specification</a>
+     *     for information about standard KDF algorithm names.
+     * @param provider
+     *     the provider to use for this key derivation
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if the specified provider does not support the specified KDF
+     *     algorithm
+     * @throws NoSuchProviderException
+     *     if the specified provider is not registered in the security provider
+     *     list
+     * @throws NullPointerException
+     *     if the {@code algorithm} or {@code provider} is {@code null}
+     */
+    public static KDF getInstance(String algorithm, String provider)
+        throws NoSuchAlgorithmException, NoSuchProviderException {
+        Objects.requireNonNull(provider, "provider may not be null");
```

Comment:
nit: may=>should?

Reply:
Same goes for other {@code Object.requireNonNull()} calls

Reply:
Why do the null check on {@code provider} here? For consistency sake, since both {@code algorithm} and {@code provider} are checked for non-null inside {@code getInstance(String, KDFParameters, String), you should either check both for non-null or leave it to the 3-arg getInstance() call.

Reply:
How about the "may" => "should" change?

Reply:
I think "cannot be null" or "must not be null" is more precise and unambiguous. The "should" and "may" words should be avoided. This comment applies to all exception messages with "may not" in the message.